src/cpr_data_access/models.py

```diff
@@ -686,3 +693,71 @@ def get_all_text_blocks(
                         output_values.append(block)
 
         return output_values
+
+    def _create_text_block_doc(self, doc: AnyDocument, block: TextBlock) -> dict[str, Any]:
+        """
+        Create a text block doc by creating a dict from the text block and its source doc metadata.
+
+        The idea is that the metadata is replicated for each text block to allow for
+        easy indexing and filtering.
+
+        :param doc: Document instance
+        :param block: TextBlock instance from the document
+        :return: Flattened document and text block as a dictionary
+        """
+        doc_metadata = doc.document_metadata.dict()
+        text = block.to_string()
+        block_type = block.type.name
+        block_id = block.text_block_id
+        flattened_doc = {
+            "text_block_id": block_id,
+            "text": text,
+            "type": block_type,
+            "language": block.language,
+            "type_confidence": block.type_confidence,
+            "page_number": block.page_number,
+            "coords": block.coords,
+            "document_id": doc.document_id,
+            "document_name": doc.document_name,
+            "document_source_url": doc.document_source_url,
+            "document_content_type": doc.document_content_type,
+            "document_md5_sum": doc.document_md5_sum,
+            "languages": doc.languages,
+            "translated": doc.translated,
+            "has_valid_text": doc.has_valid_text,
+        }
+        flattened_doc.update(doc_metadata)
+
+        if self.document_model == CPRDocument:
+            flattened_doc['document_url'] = doc.document_url
+            flattened_doc['document_cdn_object'] = doc.document_cdn_object
+            flattened_doc['document_description'] = doc.document_description
+            flattened_doc['document_slug'] = doc.document_slug
+
+        return flattened_doc
+
+    def _flatten_structure(self) -> List[dict[str, Any]]:
+        """
+        Flatten the structure of the dataset, returning a list of dictionaries
+        representing the documents and their text blocks.
+
+        The metadata is replicated for each text block to allow for easy indexing and filtering.
+
+        :return: List of flattened documents and text blocks as dictionaries
+        """
+        flattened_docs = [
+            self._create_text_block_doc(doc, block)
+            for doc in self.documents if doc.text_blocks is not None
+            for block in doc.text_blocks
+        ]
+        return flattened_docs
+
+    def to_huggingface(self) -> HFDataset:
+        """
+        Convert to a huggingface dataset to get access to the huggingface datasets API.
+
+        :return: Huggingface dataset
+        """
+        flattened_docs = self._flatten_structure()
+        hf_dataset = HFDataset.from_dict({k: [d[k] for d in flattened_docs] for k in flattened_docs[0]})
```

Comment:
This line was quite difficult to understand. I think it's worth renaming the variables `d` and `k` (e.g. `document`, `document_keys`) and assigning `document_keys = flattened_docs[0].keys()` explicitly on a separate line.